### PR TITLE
Use an external device if connected

### PR DESCRIPTION
**Pull request by manuq**

### CODE REVIEW

File: src/android_utils.py [hunk 1]
```diff
@@ -107,6 +109,52 @@ def get_home_folder():
...
+        return
+    askIntent = Intent(
+        Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION,
+        Uri.parse("package:org.endlessos.Key"),
+    )
+    get_activity().startActivity(askIntent)
+
+
+def has_all_files_access():
+    return Environment.isExternalStorageManager()
```
Comment (dylanmccall) [inline src/android_utils.py]: On my device (Lenovo Chromebook Duet, ChromeOS 100, Android 9), this throws an exception…

```
05-20 12:45:15.558  5243  5325 I python  :   File "/home/jenkins/workspace/kolibri-installer-android-pr/src/main.py", line 67, in <module>
05-20 12:45:15.559  5243  5325 I python  :   File "/home/jenkins/workspace/kolibri-installer-android-pr/src/android_utils.py", line 155, in has_all_files_access
05-20 12:45:15.559  5243  5325 I python  : AttributeError: type object 'android.os.Environment' has no attribute 'isExternalStorageManager'
```

(`Environment.isExternalStorageManager` was added in API level 30, whereas Android 9 is API level 28).

I see the device you're using has Android 11 now. I wonder if we need to do some feature probing?

Comment (manuq) [reply src/android_utils.py]: Ah good point! Yes I was on Android 9 before and then I updated this PR to make it work for Android 11.

File: src/main.py [hunk 2]
```diff
@@ -56,6 +61,22 @@ def _enable_kolibri_plugin(plugin_name: str, optional=False) -> bool:
...
+        ask_all_files_access()
+        start_time = datetime.now()
+        while not has_all_files_access() and not timeout:
+            time.sleep(0.05)
+            delta_time = datetime.now() - start_time
+            # FIXME: handle user denying access to external storage
+            # from the UI. For now just give up after 2 minutes.
+            timeout = delta_time.total_seconds() > 120
+
+    if not timeout:
```
Comment (dylanmccall) [inline src/main.py]: This is a little bit difficult to parse. Would be clearer if we move that loop waiting for access into another function such as `await_all_files_access()`.

Comment (manuq) [reply src/main.py]: Yes I think there should be UI for this. 1. a screen with the option of using a USB drive or download the collection (what Joana presented today). 2. if user opts for a USB drive the app should check if there is one connected, probably using [UsbManager](https://developer.android.com/reference/android/hardware/usb/UsbManager). And then ideally change this busy wait with a callback, if that's possible.

Comment (dylanmccall) [reply src/main.py]: Ah, I see! The `BroadcastReceiver` stuff I played with in #13 was easy enough to get going, but might be a useful reference, too: https://github.com/endlessm/kolibri-installer-android/pull/13/files#diff-df4cdc51bc9bbd15899f90c9cbca3a33eaf40f4fdb7f8924a619dd35e9e40b5bR21-R37 :)

File: src/android_utils.py [hunk 3]
```diff
@@ -107,6 +114,89 @@ def get_home_folder():
...
+
+def provision_endless_key_database(endless_key_paths):
+    if endless_key_paths is not None:
+        home_folder = get_home_folder()
+        if os.path.exists(os.path.join(home_folder, "db.sqlite3")):
+            logger.debug("EK database already exists, skipping.")
+            return
+        if not os.path.exists(home_folder):
+            os.mkdir(home_folder)
+        if Environment.isExternalStorageManager():
```
Comment (dylanmccall) [inline src/android_utils.py]: Looks like you missed a spot :)

```suggestion
        if has_all_files_access():
```

With this change, the app starts correctly for the first time on my Chromebook with Android 9, using an Endless Key device connected via a USB-C to USB-A adapter.

And the content is there. Yay!

**Review by dylanmccall — CHANGES_REQUESTED** [review]

File: src/android_utils.py (hunk 3 above)
Comment (manuq) [reply src/android_utils.py]: Ah thanks for spotting! I was trying with the Android 9 emulator but without a EK USB connected to it, so I haven't tested this part. Glad to know it even works with this change.

**Pull request merged**